petroflow/src/well_segment.py

```diff
@@ -1702,78 +1702,128 @@ def crop(self, length, step, drop_last=False, fill_value=0):
+            and returns a mapped value.
+            Defaults to `None`.
         mode : 'logs' or 'core'
-            A mode, specifying the length of computed mask. If 'logs', mask
-            lenght will match the length of well logs. If 'core', mask length
-            will match the heigth of `core_dl` in pixels. Defaults to
-            `'logs'`.
-        default : float, optional
-            Default value for mask if `src` doesn't contain information for
-            corresponding depth. Defaults to `numpy.nan`.
+            A mode, specifying which `WellSegment` attribute to use as a depth
+            indexer. If 'logs', each mask position corresponds to a depth from
+            `logs` and if 'core' — to one from `core_dl`. Defaults to 'logs'.
         dst : str, optional
-            `WellSegment` attribute to save the mask to. Defaults to `mask`.
+            Where to save the mask to. If `mode` is 'logs' then assign resulted
+            mask to `dst` column of `logs`. If `mode` is 'core` — store mask in
+            `dst` attribute of `WellSegment`. Defaults to 'mask'.
+        default : float, optional
+            Value to fill mask depths positions missing in `attr[src]`.
+            Defaults to `numpy.nan`.
+        drop : list of str, optional
+            Values from `attr[src]` to exclude (e.g. `[None, np.nan, ' ']`).
+        create_core_index : bool, optional
+            If True, save depth index in 'core' mode to 'core_index` attribute.
+            Defaults to False.
+        limit : int, optional
+            If specified, the maximum number of consecutive nan values to fill
+            both backward and forward with the closest value that is not nan.
+            In other words, if there is a gap with more than this number
+            of consecutive nans, it will only be partially filled.
+            If not specified, nan values are not filled.
 
         Returns
         -------
         self : type(self)
             Self with created mask.
         """
-        if src in self.attrs_fdtd_index:
-            self._create_mask_fdtd(src, column, mapping, mode, default, dst)
-        elif src in self.attrs_depth_index:
-            self._create_mask_depth_index(src, column, mapping, mode, default, dst)
+        if mode not in ["core", "logs"]:
+            raise ValueError('Unknown mode: ', mode)
+
+        src_series = getattr(self, attr)[src]
+
+        if not src_series.index.is_monotonic_increasing:
+            src_series = src_series.sort_index(level=0)
+        if drop is not None:
+            src_series = src_series[~src_series.isin(drop)]
+
+        src_index = src_series.index
+        src_values = src_series.values
+
+        if mapping:
+            uniques, indices = np.unique(src_series, return_inverse=True)
+            if isinstance(mapping, dict):
+                src_values = np.array([mapping[x] for x in uniques])[indices]
```

Comment:
```suggestion
            if isinstance(mapping, dict):
                mapping = lambda x, mapping=mapping: mapping.get(x, default)
```
To avoid code duplication and allow for a callable check later on instead of `elif`

Reply:
Usage of default value when mapping by dict will lead to unexpected result, when some key is forgotten in mapping scheme. It better fail with `KeyError`.

How about latest commit solution — `mapping = mapping.get if isinstance(mapping, dict) else mapping`?

Reply:
It will return `None` if the key is forgotten, using `default` here looks better.

In fact, I don't see anything wrong with a missing key in the `mapping`, you can even implement the following logic:
```
mapping = lambda x, mapping=mapping: mapping.get(x, x)
```

Reply:
Imagine having a series like `['a', 'a', …, 'a', 'b', 'a', …, 'a', 'a']`. If you wish to map it via dict to `[0, 0, …, 0, 1, 0, …, 0, 0]` and for some reason forget about the `'b'` key (i.e. use `{'a': 0}` as a mapping), than finding this error will take you some time.

Contrary, if you wish not to map some values, you can explicitly define the identity mapping through the dict.